source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -351,29 +453,78 @@ void RedisCluster::RedisDiscoverySession::onResponse(
       return;
     }
 
-    // Field 2: Primary address for slot range
-    auto primary_address = ProcessCluster(slot_range[SlotPrimary]);
-    if (!primary_address) {
+    // Row 3: Primary slot address
+    if (!validateCluster(slot_range[SlotPrimary])) {
       onUnexpectedResponse(value);
       return;
     }
+    // Try to parse primary slot address as IP address
+    // It may fail in AWS ElastiCache use case: it uses hostnames instead of IPs.
+    // If this is the case - we'll come back later and try to resolve hostnames asynchronously.
+    ClusterSlot slot(slot_range[SlotRangeStart].asInteger(), slot_range[SlotRangeEnd].asInteger(),
+                     processClusterByIP(slot_range[SlotPrimary]));
+    if (slot.primary() == nullptr) {
+      // Primary address is hostname: save the name for further resolving
+      const auto& array = slot_range[SlotPrimary].asArray();
+      slot.primary_hostname_ = array[0].asString();
+      slot.primary_port_ = array[1].asInteger();
+    }
```

Comment:
This flow is a bit confusing to follow (why are we checking `primary()`?), how about something like 

```
ClusterSlot slot(...);

if (slot.needsDnsResolution()) {
  slot.setPrimaryAddressForDnsResolution(array[0].asString(), array[1].asInteger());
}
```

Reply:
The `processClusterByIP()` method returns a `nullptr` if the address found in `slot_range` isn't of the format `IP:PORT`. So, the `slot(...)` initialization sets the `primary_` to `nullptr`, hence the check for `primary()`.

Would you recommend abstracting away the `nullptr` check against `slot.primary_` behind the `needsDnsResolution()` method?